src/snowflake/snowpark/dataframe_reader.py

```diff
@@ -1147,27 +1147,51 @@ def _generate_partition(
             )
 
         # decide stride length
-        upper_stride = processed_upper_bound / actual_num_partitions
-        lower_stride = processed_lower_bound / actual_num_partitions
-        stride = upper_stride - lower_stride
+        upper_stride = (
+            processed_upper_bound / decimal.Decimal(actual_num_partitions)
+        ).quantize(decimal.Decimal("1e-18"), rounding=ROUND_HALF_EVEN)
+        lower_stride = (
+            processed_lower_bound / decimal.Decimal(actual_num_partitions)
+        ).quantize(decimal.Decimal("1e-18"), rounding=ROUND_HALF_EVEN)
+        preciseStride = upper_stride - lower_stride
+        stride = int(preciseStride)
+
+        lost_num_of_strides = (
+            (preciseStride - decimal.Decimal(stride))
+            * decimal.Decimal(actual_num_partitions)
+            / decimal.Decimal(stride)
+        )
+        lowerBoundWithStrideAlignment = lower_bound + int(
+            (lost_num_of_strides / 2 * decimal.Decimal(stride)).quantize(
+                decimal.Decimal("1"), rounding=ROUND_HALF_UP
+            )
+        )
+
+        current_value = lowerBoundWithStrideAlignment
 
         partition_queries = []
         for i in range(actual_num_partitions):
-            left = (
-                processed_lower_bound + i * stride
-                if column_type != int
-                else int(processed_lower_bound + i * stride)
+            l_bound = (
+                f"{column} >= {self._to_external_value(current_value, column_type)}"
```

Comment:
in spark
> partitionColumn must be a numeric, date, or timestamp column

do we handle all 3 types here?

Reply:
I found I actually wrote the code to support all types, so I just added a test here